### PR TITLE
Refactor Launchers with class methods

### PR DESCRIPTION
**Pull request by mkandler**
Updates the DenLauncher and LocalLauncher classes to use classmethods for up and teardown. This allows for moving more cluster updating logic into the launchers from the OnDemandCluster

Also:
- Adds a spinner to show in logs during "info_spinner" events

### CODE REVIEW

**Comment by mkandler** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1345?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1345** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1345?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1343** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1343?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1336** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1336?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mkandler and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/resources/hardware/on_demand_cluster.py [hunk 1]
```diff
@@ -540,28 +527,11 @@ def up(self, verbose: bool = True, force: bool = False):
...
-                )
-
-            # TODO: Maybe apply the below to both cases?
-            self.restart_server()
-
-            if rns_client.autosave_resources():
-                self.save()
+            LocalLauncher.up(cluster=self, verbose=verbose)
 
         return self
```
Comment (jlewitt1) [inline runhouse/resources/hardware/on_demand_cluster.py:536]: does this approach still ensure that "self" here is updated from the `launcher.up()` call?

File: runhouse/resources/hardware/launcher_utils.py [hunk 2]
```diff
@@ -122,107 +127,119 @@ def up(self, verbose: bool = True):
...
-        super().__init__(cluster)
-        self._validate_provider()
-
-    def _validate_provider(self):
+    @staticmethod
+    def _validate_provider(cluster):
         """Check if LocalLauncher supports the provided cloud provider."""
-        supported_providers = ["cheapest"] + self.supported_providers()
-        if self.cluster.provider not in supported_providers:
+        supported_providers = ["cheapest"] + supported_providers()
```
Comment (jlewitt1) [inline runhouse/resources/hardware/launcher_utils.py:186]: prob makes sense for this to be a class method 

```suggestion
        supported_providers = ["cheapest"] + cls.supported_providers()
```

**Comment by jlewitt1** [comment]
Maybe we consolidate the logic from the other PRs in the stack into this one? There's a bunch of irrelevant code now that has more to do with the old approach of using a separate logs endpoint

File: runhouse/resources/hardware/on_demand_cluster.py (hunk 1 above)
Comment (mkandler) [reply runhouse/resources/hardware/on_demand_cluster.py:536]: Yup! As long as the cluster reference is maintained in the method, the changes pass through to `self`!
Took a bit of experimentation but definitely working

File: runhouse/resources/hardware/launcher_utils.py [hunk 3]
```diff
@@ -232,11 +260,12 @@ def up(self, verbose: bool = True):
...
             raise e
 
-    def teardown(self, verbose: bool = True):
+    @classmethod
+    def teardown(cls, cluster, verbose: bool = True):
         """Tearing down a cluster locally via Sky."""
         import sky
 
-        sky.down(self.cluster.name)
+        sky.down(cluster.name)
```
Comment (jlewitt1) [inline runhouse/resources/hardware/launcher_utils.py:269]: maybe we keep a similar structure to `up`, where for the den launcher we handle the saving in Den, and for the local launcher we add this after calling `sky down`: 

```
        # Save to Den with updated null IPs
        if rns_client.autosave_resources():
            self.save()
```

**Pull request merged**